In [73]:
#1. Import the necessary libraries.
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [74]:
customer_df = pd.read_csv('files_for_lab/we_fn_use_c_marketing_customer_value_analysis.csv')
customer_df.head()

Customer       State  Customer Lifetime Value Response  Coverage Education  \
0  BU79786  Washington              2763.519279       No     Basic  Bachelor   
1  QZ44356     Arizona              6979.535903       No  Extended  Bachelor   
2  AI49188      Nevada             12887.431650       No   Premium  Bachelor   
3  WW63253  California              7645.861827       No     Basic  Bachelor   
4  HB64268  Washington              2813.692575       No     Basic  Bachelor   

  Effective To Date EmploymentStatus Gender  Income  ...  \
0           2/24/11         Employed      F   56274  ...   
1           1/31/11       Unemployed      F       0  ...   
2           2/19/11         Employed      F   48767  ...   
3           1/20/11       Unemployed      M       0  ...   
4            2/3/11         Employed      M   43836  ...   

  Months Since Policy Inception Number of Open Complaints  Number of Policies  \
0                             5                         0                   1   
1                            42                         0                   8   
2                            38                         0                   2   
3                            65                         0                   7   
4                            44                         0                   1   

      Policy Type        Policy  Renew Offer Type  Sales Channel  \
0  Corporate Auto  Corporate L3            Offer1          Agent   
1   Personal Auto   Personal L3            Offer3          Agent   
2   Personal Auto   Personal L3            Offer1          Agent   
3  Corporate Auto  Corporate L2            Offer1    Call Center   
4   Personal Auto   Personal L1            Offer1          Agent   

  Total Claim Amount  Vehicle Class Vehicle Size  
0         384.811147   Two-Door Car      Medsize  
1        1131.464935  Four-Door Car      Medsize  
2         566.472247   Two-Door Car      Medsize  
3         529.881344            SUV      Medsize  
4         138.130879  Four-Door Car      Medsize  

[5 rows x 24 columns]

In [75]:
#2. Find all of the categorical data. Save it in a categorical_df variable.
categorical_df=customer_df.select_dtypes(object)

In [76]:
#3. Check for NaN values.
categorical_df.isna().sum()/len(categorical_df)

Customer             0.0
State                0.0
Response             0.0
Coverage             0.0
Education            0.0
Effective To Date    0.0
EmploymentStatus     0.0
Gender               0.0
Location Code        0.0
Marital Status       0.0
Policy Type          0.0
Policy               0.0
Renew Offer Type     0.0
Sales Channel        0.0
Vehicle Class        0.0
Vehicle Size         0.0
dtype: float64

In [77]:
# 4. Check all unique values of columns.
# (We effectively excluded Customer and Effective To Date columns as they were ID column and Date column respectively.)
for col in categorical_df.columns:
    if col!='Customer' and col!='Effective To Date':
        print(categorical_df[col].value_counts())
        print()

California    3150
Oregon        2601
Arizona       1703
Nevada         882
Washington     798
Name: State, dtype: int64

No     7826
Yes    1308
Name: Response, dtype: int64

Basic       5568
Extended    2742
Premium      824
Name: Coverage, dtype: int64

Bachelor                2748
College                 2681
High School or Below    2622
Master                   741
Doctor                   342
Name: Education, dtype: int64

Employed         5698
Unemployed       2317
Medical Leave     432
Disabled          405
Retired           282
Name: EmploymentStatus, dtype: int64

F    4658
M    4476
Name: Gender, dtype: int64

Suburban    5779
Rural       1773
Urban       1582
Name: Location Code, dtype: int64

Married     5298
Single      2467
Divorced    1369
Name: Marital Status, dtype: int64

Personal Auto     6788
Corporate Auto    1968
Special Auto       378
Name: Policy Type, dtype: int64

Personal L3     3426
Personal L2     2122
Personal L1     1240
Corporate L3    1014
Corporate L2

In [78]:
#5. Check dtypes. Do they all make sense as categorical data?

Excluding Customer and Effective to Date, the values in other columns are distinct and countable as expected from a categorical column and none of the column values are specified as integer within the categorical column. 

The Customer column is an ID column containing alpha numeric and Effective to Date is effectively a Date column and needs to converted to one.

Other categorical column have scope for revision and transformation to other kinds (like ordinal etc.)

In [79]:
categorical_df['Effective To Date'] =  pd.to_datetime(categorical_df['Effective To Date'])
categorical_df.head()

Customer       State Response  Coverage Education Effective To Date  \
0  BU79786  Washington       No     Basic  Bachelor        2011-02-24   
1  QZ44356     Arizona       No  Extended  Bachelor        2011-01-31   
2  AI49188      Nevada       No   Premium  Bachelor        2011-02-19   
3  WW63253  California       No     Basic  Bachelor        2011-01-20   
4  HB64268  Washington       No     Basic  Bachelor        2011-02-03   

  EmploymentStatus Gender Location Code Marital Status     Policy Type  \
0         Employed      F      Suburban        Married  Corporate Auto   
1       Unemployed      F      Suburban         Single   Personal Auto   
2         Employed      F      Suburban        Married   Personal Auto   
3       Unemployed      M      Suburban        Married  Corporate Auto   
4         Employed      M         Rural         Single   Personal Auto   

         Policy Renew Offer Type Sales Channel  Vehicle Class Vehicle Size  
0  Corporate L3           Offer1         Agent   Two-Door Car      Medsize  
1   Personal L3           Offer3         Agent  Four-Door Car      Medsize  
2   Personal L3           Offer1         Agent   Two-Door Car      Medsize  
3  Corporate L2           Offer1   Call Center            SUV      Medsize  
4   Personal L1           Offer1         Agent  Four-Door Car      Medsize

In [80]:
#6. Does any column contain alpha and numeric data? Decide how to clean it.

The column policy and offer contains alpha and numeric data. On observation, it is found that the numeric values in these columns are nothing but the levels, and string values seem redundant. The numerical values can be converted to ordinal values through transformation. (The column policy will be converted later in #8.)

In [81]:
categorical_df['Renew Offer Type']=categorical_df['Renew Offer Type'].apply(lambda x: x[-1])
categorical_df.head()

Customer       State Response  Coverage Education Effective To Date  \
0  BU79786  Washington       No     Basic  Bachelor        2011-02-24   
1  QZ44356     Arizona       No  Extended  Bachelor        2011-01-31   
2  AI49188      Nevada       No   Premium  Bachelor        2011-02-19   
3  WW63253  California       No     Basic  Bachelor        2011-01-20   
4  HB64268  Washington       No     Basic  Bachelor        2011-02-03   

  EmploymentStatus Gender Location Code Marital Status     Policy Type  \
0         Employed      F      Suburban        Married  Corporate Auto   
1       Unemployed      F      Suburban         Single   Personal Auto   
2         Employed      F      Suburban        Married   Personal Auto   
3       Unemployed      M      Suburban        Married  Corporate Auto   
4         Employed      M         Rural         Single   Personal Auto   

         Policy Renew Offer Type Sales Channel  Vehicle Class Vehicle Size  
0  Corporate L3                1         Agent   Two-Door Car      Medsize  
1   Personal L3                3         Agent  Four-Door Car      Medsize  
2   Personal L3                1         Agent   Two-Door Car      Medsize  
3  Corporate L2                1   Call Center            SUV      Medsize  
4   Personal L1                1         Agent  Four-Door Car      Medsize

In [82]:
#7. Would you choose to do anything else to clean or wrangle the categorical data? Comment your decisions.

State - They are countable to 5 unique values, so we wouldn't do anything to this column, until we start observing that more state values start pooling in this column and count of those state values remain less than 1000, we can effectively group them to a value called Others.

Policy and Offer - As explained in the above question, they will need to be converted to ordinal values.

Vehicle Class - The column values can be combined to a fewer distinct values by grouping similar class/segment of vehicles.


In [83]:
#8. Compare policy_type and policy. What information is contained in these columns. Can you identify what is important?

The column policy contains redundant data and the similar information is already available in the column policy type. The two columns are tighly correlated to each other. That is, the only value expected in the policy column is the policy level. The column name policy can later be renamed as policy level.

Before making the change to column Policy, we make sure the 2 column values are indeed related to each other (although it's obvious). That is we make sure that the common value in both the columns (values between Personal, Corporate, Special) is equal.

In [84]:
if (categorical_df['Policy Type'].apply(lambda x: x.split(" ")[0])==categorical_df['Policy'].apply(lambda x: x.split(" ")[0])).all():
    categorical_df['Policy']=categorical_df['Policy'].apply(lambda x: x[-1])

categorical_df.rename(columns={'Policy': 'Policy Level'}, inplace=True)
categorical_df.head()

Customer       State Response  Coverage Education Effective To Date  \
0  BU79786  Washington       No     Basic  Bachelor        2011-02-24   
1  QZ44356     Arizona       No  Extended  Bachelor        2011-01-31   
2  AI49188      Nevada       No   Premium  Bachelor        2011-02-19   
3  WW63253  California       No     Basic  Bachelor        2011-01-20   
4  HB64268  Washington       No     Basic  Bachelor        2011-02-03   

  EmploymentStatus Gender Location Code Marital Status     Policy Type  \
0         Employed      F      Suburban        Married  Corporate Auto   
1       Unemployed      F      Suburban         Single   Personal Auto   
2         Employed      F      Suburban        Married   Personal Auto   
3       Unemployed      M      Suburban        Married  Corporate Auto   
4         Employed      M         Rural         Single   Personal Auto   

  Policy Level Renew Offer Type Sales Channel  Vehicle Class Vehicle Size  
0            3                1         Agent   Two-Door Car      Medsize  
1            3                3         Agent  Four-Door Car      Medsize  
2            3                1         Agent   Two-Door Car      Medsize  
3            2                1   Call Center            SUV      Medsize  
4            1                1         Agent  Four-Door Car      Medsize

In [85]:
#9. Check number of unique values in each column, can they be combined in any way to ease encoding? Comment your thoughts and make those changes.

In [86]:
def standardize_vehicle_class(x):
    if x in ['Two-Door Car', 'Four-Door Car']:
        return 'Standard'
    elif x in ['SUV', 'Sports Car']:
        return 'Sports'
    else:
        return 'Luxury'
    
categorical_df['Vehicle Class']=categorical_df['Vehicle Class'].apply(standardize_vehicle_class)
categorical_df['Vehicle Class'].value_counts()

Standard    6507
Sports      2280
Luxury       347
Name: Vehicle Class, dtype: int64

Education and Employment Status can be also grouped individually within their own values but prefer to leave them as such as they would lose the value otherwise which their distinct values otherwise reflect. Explained further below.

+ Bachelor                2748
+ College                 2681
+ High School or Below    2622
+ Master                   741
+ Doctor                   342

Bachelor and College look similar but it it interpreted differently in in different countries, so prefer to leave them untouched.

+ Employed         5698
+ Unemployed       2317
+ Medical Leave     432
+ Disabled          405
+ Retired           282

Prefer to leave Retired ungrouped as it is expressed different from Employed and Unemployed. Disabled and Medical Leave could be grouped together but again they have their own meaning and could lose their individual meaning, if grouped.